### PR TITLE
Cloning Rework

### PR DESCRIPTION
**Pull request by Elijahrane**
## About the PR 
Yeah I revived the old fishfish pr, draft because I still have a lot of cleanup and I can't merge upstream because it crashes with a weightless bug.

**Changelog**

:cl: Fishfish458, Rane
- tweak: Reworked cloning. It now takes a cloning pod, cloning console, and medical scanner linked together with a multitool.
- remove: You can no longer prescan to be cloned. Your body has to be in the scanner at the time of cloning.

### CODE REVIEW

**Comment by Elijahrane** [comment]
What do you want me to do about body reassemble since it's already pretty broken and incompatible with this

**Comment by EmoGarbage404** [comment]
> What do you want me to do about body reassemble since it's already pretty broken and incompatible with this

Um you can probably leave it commented out and then i can fix it after the pr is merged. On thing of note, im not sure the species field in humanoid appearance is properly updated and tracked.

**Comment by EmoGarbage404** [comment]
Ive made fixes to skeletons in a new pr, so i suggest removing all unnecessary changes in your code lest you may run into a lot of conflicts.

**Comment by Elijahrane** [comment]
I'll just fix it after that's merged because this isn't gonna build if I uncomment that

File: Content.Server/Body/Components/BodyReassembleComponent.cs [hunk 1]
```diff
@@ -1,35 +1,34 @@
...
-        public InstantAction? ReassembleAction = null;
+//         [DataField("action")]
+//         public InstantAction? ReassembleAction = null;
 
-        public CancellationTokenSource? CancelToken = null;
-    }
-}
+//         public CancellationTokenSource? CancelToken = null;
+//     }
+// }
```
Comment (mirrorcult) [inline Content.Server/Body/Components/BodyReassembleComponent.cs:34]: has this been discussed with emogarbage at all as for whether this is getting fixed soon, cause ideally this is uncommented before merge

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        private void OnPowerChanged(EntityUid uid, CloningConsoleComponent component, PowerChangedEvent args)
+        {
+            component.Powered = args.Powered;
+            UpdateUserInterface(component);
+        }
+
+        private void OnNewLink(EntityUid uid, CloningConsoleComponent component, NewLinkEvent args)
+        {
+            Logger.Error("args.TransmitterPort is " + args.TransmitterPort);
+            Logger.Error("args.Receiver is: " + args.Receiver);
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:73]:
```suggestion
```

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            if (TryComp<MedicalScannerComponent>(args.Receiver, out var scanner) && args.TransmitterPort == "MedicalScannerSender")
+            {
+                Logger.Error("Adding scanner...");
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:76]:
```suggestion
```

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                component.GeneticScanner = args.Receiver;
+                scanner.ConnectedConsole = uid;
+            }
+
+            if (TryComp<CloningPodComponent>(args.Receiver, out var pod) && args.TransmitterPort == "CloningPodSender")
+            {
+                Logger.Error("Adding pod...");
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:83]:
```suggestion
```

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+            if (scanner != null)
+            {
+                Transform((EntityUid) scanner).Coordinates.TryDistance(EntityManager, Transform((console)).Coordinates, out float scannerDistance);
+                consoleComp.GeneticScannerInRange = scannerDistance <= consoleComp.MaxDistance;
+            }
+            if (cloningPod != null)
+            {
+                Transform((EntityUid) cloningPod).Coordinates.TryDistance(EntityManager, Transform((console)).Coordinates, out float podDistance);
+                consoleComp.CloningPodInRange = podDistance <= consoleComp.MaxDistance;
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:175]:
```suggestion
                Transform(scanner.Value).Coordinates.TryDistance(EntityManager, Transform((console)).Coordinates, out float scannerDistance);
                consoleComp.GeneticScannerInRange = scannerDistance <= consoleComp.MaxDistance;
            }
            if (cloningPod != null)
            {
                Transform(cloningPod.Value).Coordinates.TryDistance(EntityManager, Transform((console)).Coordinates, out float podDistance);
                consoleComp.CloningPodInRange = podDistance <= consoleComp.MaxDistance;
```

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            }
+
+            UpdateUserInterface(consoleComp);
+        }
+        private CloningConsoleBoundUserInterfaceState GetUserInterfaceState(CloningConsoleComponent consoleComponent)
+        {
+            ClonerStatus clonerStatus = ClonerStatus.Ready;
+
+            // genetic scanner info
+            string scanBodyInfo = "Unknown";
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:185]: localize

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            }
+            RecheckConnections(uid, component.CloningPod, component.GeneticScanner, component);
+        }
+
+        private void OnPortDisconnected(EntityUid uid, CloningConsoleComponent component, PortDisconnectedEvent args)
+        {
+            if (args.Port == "MedicalScannerSender")
+                component.GeneticScanner = null;
+
+            if (args.Port == "CloningPodSender")
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:95]: these should be static strings/consts somewhere

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                        if (mindComp == null || mindComp.Mind == null || mindComp.Mind.UserId == null || !_playerManager.TryGetSessionById(mindComp.Mind.UserId.Value, out var client))
+                        {
+                            clonerStatus = ClonerStatus.NoMindDetected;
+                        }
+                    }
+                }
+            }
+
+            // cloning pod info
+            var cloneBodyInfo = "Unknown";
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:217]: loc

File: Content.Server/Cloning/CloningConsoleSystem.cs [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+            float cloningProgress = 0;
+            float cloningTime = 30f;
+            bool clonerProgressing = false;
+            bool clonerConnected = false;
+            bool clonerMindPresent = false;
+            bool clonerInRange = consoleComponent.CloningPodInRange;
+            if (consoleComponent.CloningPod != null && TryComp<CloningPodComponent>(consoleComponent.CloningPod, out var clonePod)
+            && Transform((EntityUid) consoleComponent.CloningPod).Anchored)
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningConsoleSystem.cs:225]:
```suggestion
            && Transform(consoleComponent.CloningPod.Value).Anchored)
```
etc for all other cases of that

File: Content.Server/Cloning/CloningSystem.cs [hunk 10]
```diff
@@ -39,114 +65,132 @@ internal void TransferMindToClone(Mind.Mind mind)
...
+        public override void Update(float frameTime)
         {
-            return MindToId.ContainsKey(mind);
+            foreach (var cloning in EntityManager.EntityQuery<CloningPodComponent>())
+            {
+                if (!IsPowered(cloning))
+                    continue;
+
+                if (cloning.BodyContainer.ContainedEntity != null)
+                {
```
Comment (mirrorcult) [inline Content.Server/Cloning/CloningSystem.cs:164]: this should really use an `ActiveCloningPodComponent` for perf

File: Content.Server/Cloning/Components/CloningConsoleComponent.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+        /// Maximum distance between console and one if its machines
+        [DataField("maxDistance")]
+        public float MaxDistance = 4f;
+
+        public bool GeneticScannerInRange = true;
+
+        public bool CloningPodInRange = true;
+
+        public bool Powered = false;
```
Comment (mirrorcult) [inline Content.Server/Cloning/Components/CloningConsoleComponent.cs:35]: this doesnt need to be stateful

File: Content.Shared/Cloning/SharedCloningConsole.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
+﻿using Robust.Shared.Serialization;
+
+namespace Content.Shared.Cloning.CloningConsole
+{
+        [Serializable, NetSerializable]
+        public sealed class CloningConsoleBoundUserInterfaceState : BoundUserInterfaceState
+        {
+            public readonly string? ScannerBodyInfo;
+            public readonly string? ClonerBodyInfo;
```
Comment (mirrorcult) [inline Content.Shared/Cloning/SharedCloningConsole.cs:9]: entire file is indented one too many

**Review by mirrorcult — CHANGES_REQUESTED** [review]
will need another review + i need to test ingame as well

**Comment by Elijahrane** [comment]
image

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**